client/db/benches/state_access.rs

```diff
@@ -159,152 +204,191 @@ fn generate_storage() -> (Vec<Vec<u8>>, Vec<(Vec<u8>, Vec<u8>)>) {
 
-	bench_single_value(
-		BenchmarkConfig::TrieNodeCache,
-		"with trie node cache and reading 4 times each key in a row",
-		4,
-	);
-	bench_single_value(
-		BenchmarkConfig::NoCache,
-		"no cache and reading 4 times each key in a row",
-		4,
-	);
+	// Read a single value once. There should be no difference here between the cache and no cache
+	// variants.
+	let mut group = c.benchmark_group("Read a single value once");
+	group.sample_size(20);
+	group.measurement_time(Duration::from_secs(30));
+	run_benchmark_with_and_without_cache(group, &all_keys[..1], 1, 1, AccessType::Storage);
 
-	group.finish();
+	// Read a single value many times, with an empty shared cache but populated local cache (except
+	// the first time).
+	let mut group = c.benchmark_group("Read a single value 1024 times (local cache only)");
+	group.sample_size(20);
+	group.measurement_time(Duration::from_secs(30));
+	run_benchmark_with_and_without_cache(group, &all_keys[..1], 1, 1024, AccessType::Storage);
 
-	let mut group = c.benchmark_group("Hashing a value");
+	// Read a single value many times, with a populated shared cache but empty local cache (except
+	// the first time).
+	let mut group = c.benchmark_group("Read a single value 1024 times (shared cache only)");
+	group.sample_size(20);
+	group.measurement_time(Duration::from_secs(30));
+	run_benchmark_with_and_without_cache(group, &all_keys[..1], 1024, 1, AccessType::Storage);
 
-	let mut bench_single_value = |config, desc, multiplier| {
-		let backend = create_backend(config, &path);
+	// Read a single value many times, with a populated shared cache and populated local cache
+	// (except the first time).
+	let mut group = c.benchmark_group("Read a single value 1024 times (both caches)");
+	group.sample_size(20);
+	group.measurement_time(Duration::from_secs(30));
+	run_benchmark_with_and_without_cache(group, &all_keys[..1], 32, 32, AccessType::Storage);
 
-		group.bench_function(desc, |b| {
-			b.iter_batched(
-				|| backend.state_at(block_hash).expect("Creates state"),
-				|state| {
-					for key in keys.iter().take(1).cycle().take(multiplier) {
-						let _ = state.storage_hash(&key).expect("Doesn't fail").unwrap();
-					}
-				},
-				BatchSize::SmallInput,
-			)
-		});
-	};
+	let mut group = c.benchmark_group("Hash a value once");
+	group.sample_size(20);
+	group.measurement_time(Duration::from_secs(30));
+	run_benchmark_with_and_without_cache(group, &all_keys[..1], 1, 1, AccessType::StorageHash);
 
-	bench_single_value(
-		BenchmarkConfig::TrieNodeCache,
-		"with trie node cache and hashing the key once",
-		1,
-	);
-	bench_single_value(BenchmarkConfig::NoCache, "no cache and hashing the key once", 1);
+	let mut group = c.benchmark_group("Hash a value 1024 times");
+	group.sample_size(20);
+	group.measurement_time(Duration::from_secs(30));
+	run_benchmark_with_and_without_cache(group, &all_keys[..1], 1, 1024, AccessType::StorageHash);
```

Comment:
Maybe we could test with different "cache setups" as you do it for a single value above?